spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,371 @@
+    }
+
+    // create prefixed key for proof verification
+    prefixedKey := abortTransactionUnless(commitmenttypes.ApplyPrefix(prefix, commitmenttypes.NewMerklePath(key)))
+ 
+    // verify each consensus state and connection state starting going from Z --> A
+    // finally verify the keyproof on A within B's verified view of A's consensus state.
+    abortTransactionUnless(VerifyMultiHopProofMembership(consensusState, &proofs, &prefixedKey, value))
+}
+
+// VerifyMultiHopProofMembership verifies a multihop membership proof including all intermediate state proofs.
+func VerifyMultiHopProofMembership(
+    consensusState exported.ConsensusState,
+    proofs *MsgMultihopProofs,
+    prefixedKey commitmenttypes.MerklePath,
+    value []byte,
+) {
+    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
+    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))
+ 
+    abortTransactionUnless(VerifyMultiHopConsensusStateProof(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))
+
+    keyProof := abortTransactionUnless(Unmarshal(proofs.KeyProof.Proof))
+    
+    // the consensus state of the source chain (chain[0] on chain[1])
+    consState := abortTransactionUnless(UnmarshalInterface(proofs.ConsensusProofs[0].Value))
+
+    abortTransactionUnless(keyProof.VerifyMembership(
+        commitmenttypes.GetSDKSpecs(),
+        consState.GetRoot(),
+        prefixedKey,
+        value,
+    ))
+}
+
+// VerifyMultiHopConsensusStateProof verifies the consensus state of each consecutive consensus/connection state starting
+// from chain[N-1] on the destination (chain[N]) and finally proving the source chain consensus/connection state.
+func VerifyMultiHopConsensusStateProof(
+ consensusState exported.ConsensusState,
+ consensusProofs []*MultihopProof,
+ connectionProofs []*MultihopProof,
+) {
+    // reverse iterate through proofs to prove from destination to source
+    for i := len(consensusProofs) - 1; i >= 0; i-- {
+        consStateProof := consensusProofs[i]
+        connectionProof := connectionProofs[i]
+        conState := abortTransactionUnless(UnmarshalInterface(consStateProof.Value))
+
+        // prove the consensus state of chain[i] in chain[i+1]
+        consensusProof := abortTransactionUnless(Unmarshal(consStateProof.Proof))
+        abortTransactionUnless(consensusProof.VerifyMembership(
+            commitmenttypes.GetSDKSpecs(),
+            consensusState.GetRoot(),
+            *consStateProof.PrefixedKey,
+            consStateProof.Value,
+        ))
+
+        // prove the connection state of chain[i] in chain[i+1]
+        connectionProof := abortTransactionUnless(Unmarshal(connectionProof.Proof))
+        abortTransactionUnless(connectionProof.VerifyMembership(
+            commitmenttypes.GetSDKSpecs(),
+            consensusState.GetRoot(),
+            *connectionProof.PrefixedKey,
+            connectionProof.Value, // this should be from connectionHops
```

Comment:
Not clear to me, do we check anywhere that the consensus states are from clients associated with the actual connections?

Reply:
So, we check each connection is OPEN and matching the connections specified in the channel connectionHops field [here](https://github.com/polymerdao/ibc/blame/bo/ics-32/spec/core/ics-033-multi-hop/README.md#L276).

We also check that each of those provided connectionEnds are within the provided consensus states [here](https://github.com/polymerdao/ibc/blame/bo/ics-32/spec/core/ics-033-multi-hop/README.md#L311)

At a hight level we prove:
- The connectionHops used to establish the channel match the connections provided in the multihop proof and are OPEN
- the connectionEnds are proven within the provided consensus states
- the consensus states are proven recursively starting from the trusted consensus state on the receiving chain.

Does that suffice?

Reply:
Two clients may have the same consensus state. So I was looking for an explicit check that `connectionEnd.ClientId` is the same as the `clientId` for which the consensus proof was collected, i.e. same as the ID in its key path. Check would be something like:
```
consensusProofClientId = strings.Split(consensusProof.PrefixedKey.GetKeyPath()[len(consensusProof.PrefixedKey.KeyPath)-1], "/")..
abortTransactionUnless(connectionEnd.ClientId == consensusProofClientId)
```

Reply:
Yeah, I think we will likely need to bundle clientStates with the multihop proof as well per offline discussion. In this case we should definitely verify the clientid of the clientstate matches the consensus/connection state.

Reply:
@ancazamfir also, FYI, I'm not sure if you saw this. Here is the offline doc summarizing the review comments: https://docs.google.com/document/d/1lFxGZwapWANirgCO3NthsDGHgmBdSo_71alj14rAhnU/edit#

Reply:
Feel free to drop comments there and we can enshrine the decision in GH.

Reply:
thanks, i will take a look and provide comments there

Reply:
To summarize the offline discussion and google doc collaboration:

- delayPeriod will be enforced by computing the maximum delayPeriod across all connections in the channel path and enforcing that delayPeriod on the final consensusState on the verifier. This should suffice because by construction all recursively embedded consensus states must have existed for at least as long as the final consensus state.
- Client states are NOT required to be bundled with the multi-hop proof. Instead a new chanCloseFrozen() method will be introduced to allow relayers to post a multi-hop proof of a frozen client state starting from the misbehaving chain in the channel path. The proof is not symmetric, but relayers can send a unique proof (representing the hops from each end to the misbehaving chain) to actively freeze the channel.
- Frozen channels can not process packets and a new FROZEN state could be introduced to help distinguish this situation.
- If a client is frozen in the channel path then it would not be possible to send newly committed packets as the consensus state associated with the frozen client would not be updatable making it impossible to create new multi-hop proofs.

Hopefully I represented that correctly and didn't miss anything.